middleware/jwt/src/middleware.rs

```diff
@@ -0,0 +1,333 @@
+/// extern crate serde;
+/// #[macro_use]
+/// extern crate serde_derive;
+///
+/// use futures::future;
+/// use gotham::{
+///     helpers::http::response::create_empty_response,
+///     handler::HandlerFuture,
+///     pipeline::{
+///         new_pipeline,
+///         set::{finalize_pipeline_set, new_pipeline_set},
+///     },
+///     router::{builder::*, Router},
+///     state::{State, FromState},
+/// };
+/// use gotham_middleware_jwt::{JWTMiddleware, AuthorizationToken};
+/// use hyper::{Response, StatusCode};
+///
+/// #[derive(Deserialize, Debug)]
+/// struct Claims {
+///     sub: String,
+///     exp: usize,
+/// }
+///
+/// fn handler(state: State) -> Box<HandlerFuture> {
+///     {
+///         let token = AuthorizationToken::<Claims>::borrow_from(&state);
+///         // auth.token -> TokenData
+///     }
+///     let res = create_empty_response(&state, StatusCode::OK);
+///     Box::new(future::ok((state, res)))
+/// }
+///
+/// fn router() -> Router {
+///     let pipelines = new_pipeline_set();
+///     let (pipelines, defaults) = pipelines.add(
+///         new_pipeline()
+///             .add(JWTMiddleware::<Claims>::new("secret".as_ref()))
+///             .build(),
+///     );
+///     let default_chain = (defaults, ());
+///     let pipeline_set = finalize_pipeline_set(pipelines);
+///     build_router(default_chain, pipeline_set, |route| {
+///         route.get("/").to(handler);
+///     })
+/// }
+///
+/// # fn main() {
+/// #    let _ = router();
+/// # }
+/// ```
+pub struct JWTMiddleware<T> {
+    secret: &'static str,
+    validation: Validation,
+    claims: PhantomData<T>,
+}
+
+impl<T> JWTMiddleware<T>
+where
+    T: for<'de> Deserialize<'de> + Send + Sync,
+{
+    /// Creates a JWTMiddleware instance from the provided secret,
+    /// which, by default, uses HS256 as the crypto scheme.
+    pub fn new(secret: &'static str) -> Self {
+        let validation = Validation::default();
+
+        JWTMiddleware {
+            secret,
+            validation,
+            claims: PhantomData,
+        }
+    }
+
+    /// Create a new instance of the middleware by appending new
+    /// validation constraints.
+    pub fn validation(self, validation: Validation) -> Self {
+        JWTMiddleware { validation, ..self }
+    }
+}
+
+impl<T> Middleware for JWTMiddleware<T>
+where
+    T: for<'de> Deserialize<'de> + Send + Sync + 'static,
+{
+    fn call<Chain>(self, mut state: State, chain: Chain) -> Box<HandlerFuture>
+    where
+        Chain: FnOnce(State) -> Box<HandlerFuture>,
+    {
+        trace!("[{}] pre-chain authentication", request_id(&state));
+
+        let token: Option<String> = {
+            let header = HeaderMap::borrow_from(&state).get(AUTHORIZATION);
+
+            match header {
+                Some(h) => match h.to_str() {
+                    Ok(hx) => {
+                        let parts: Vec<&str> = hx.rsplit(": ").collect();
+                        Some(parts[0].to_owned())
```

Comment:
It seems like we can avoid the `collect` here.

From what I can tell, this is basically doing one of those things where we split `Authorization: TOKEN`. If that's the case, can't we simply skip the first N bytes of the input and just reference the TOKEN directly? That makes this avoid allocations completely.

Reply:
Allocation was dropped in the latest version I pushed. Good point.